browser/src/Renderer/WebGLRenderer/TextRenderer/LigatureGrouper/OpenTypeLigatureGrouper.ts

```diff
@@ -0,0 +1,112 @@
+import fontManager from "font-manager"
+import fontkit, { Font } from "fontkit"
+import * as fs from "fs"
+import * as groupBy from "lodash/groupBy"
+import * as Log from "oni-core-logging"
+
+import { GlyphInfo } from "./GlyphInfo"
+import { GlyphSubstitutor } from "./GlyphSubstitutor"
+import { ILigatureGrouper } from "./ILigatureGrouper"
+
+const ligatureFeatures = ["calt", "rclt", "liga", "dlig", "clig"]
+
+export class OpenTypeLigatureGrouper implements ILigatureGrouper {
+    private readonly _font = loadFont(this._fontFamily)
+    private readonly _fontHasLigatures = this._font && checkIfFontHasLigatures(this._font)
+    private readonly _glyphSubstitutor = this._fontHasLigatures
+        ? new GlyphSubstitutor(this._font)
+        : null
+    private readonly _cache = new Map<string, string[]>()
+
+    constructor(private _fontFamily: string) {}
+
+    public getLigatureGroups(characters: string[]) {
+        if (!this._fontHasLigatures) {
+            return characters
+        }
+
+        const concatenatedCharacters = characters.join("")
+
+        const cachedLigatureGroups = this._cache.get(concatenatedCharacters)
+        if (cachedLigatureGroups) {
+            return cachedLigatureGroups
+        }
+
+        const fontGlyphs = this._font.glyphsForString(concatenatedCharacters)
+        const glyphInfos = fontGlyphs.map(
+            glyph => new GlyphInfo(this._font, glyph.id, [...glyph.codePoints], ligatureFeatures),
+        )
+        // Apply ligatures and store contextGroup metadata in the GlyphInfos wherever they applied
+        this._glyphSubstitutor.applyFeatures(ligatureFeatures, glyphInfos)
+
+        // Group GlyphInfo[] by contextGroup
+        const contextGroupDictionary = groupBy(glyphInfos, glyphInfo => glyphInfo.contextGroup)
+        const contextGroupSymbols = Object.getOwnPropertySymbols(contextGroupDictionary)
+        // TODO remove the "as any" once we upgraded to TS 3
+        const contextGroups = contextGroupSymbols.map(
+            contextGroupSymbol => contextGroupDictionary[contextGroupSymbol as any],
+        )
+
+        // Map GlyphInfo[][] to string[]
+        const ligatureGroups = contextGroups.map(glyphsInContextGroup =>
```

Comment:
Why do you need these contextGroups? fontkit already groups glyphs together and stores the grouped codePoints, so you should just be able to create your ligatureGroups based on that right? When would you have multiple glyphs grouped into a single contextGroup?

```javascript
const ligatureGroups = glyphInfos.map(g => String.fromCodePoint(...glyph.codePoints));
```

Reply:
Thanks for the review!

The new `contextGroup` property is there mainly for tracking contextual alternates from `calt` etc. where the glyphs are replaced one by one without combining them. Whenever this happens, we set the context group of all involved glyphs to be the same so that we can track which glyphs are involved in such a replacement.

I realize that the approach that we take here is a bit unconventional, but let me explain the reasoning behind it:

We have a WebGL-based editor renderer that we use for performance reasons. It uses a texture atlas approach which means every glyph is rendered into a texture exactly once and then copied over into the editor view at the required positions using the GPU. Normally, this is done character by character. To support ligatures though, we need to group the characters that will produce ligatures and have chrome render them into the atlas as one string. This is why we need the `contextGroup` approach: we need to map the glyphs back to one string as a group but still keep all other characters that don't produce ligatures separate in their own string.

We decided not to use fontkit's own rendering algorithm because it does not support auto-hinting and also no subpixel antialiasing if I remember correctly. Since both features are very important for creating a crisp and well-readable font (even more so on low-DPI screens), we decided to stick to Chrome's harfbuzz-based renderer.

Reply:
Ah I see: you need to preserve the context in order for Chrome to reproduce the correct glyphs. Yeah fontkit is more meant for layout rather than rendering so its rendering features are somewhat limited. It would be really nice if you could pass glyph ids to the canvas fillText command instead of characters... that would solve this problem and a few others I bet.

Theoretically you could re-encode the font using a different cmap that mapped characters directly to glyphs using the same ids, and then call fillText but using glyph ids as character codes. Not sure how much work that would be.

Reply:
I totally agree, being able to render by glyph ID would save us a lot of trouble :)

The re-encoding of the font file sounds interesting. I'll think about that a little more. Thanks!